Implement support for UCA collation operations

adamretter opened the pull request:
Closes https://github.com/eXist-db/exist/issues/1569

I hope this is the first step in eXist having serious support for Collation operations in XQuery 3+.

I have switched us from `java.text.Collator` to ICU's Collator, as to be able to support UCA well, we need much finer control over the Collator. Also ICU is supposed to have better performance and accuracy that the JDK equivalent.

We don't yet support the `fallback=no`, as it is not clear to me how to do that with ICU yet. However, we certainly support a huge ammount more than we did before.

@joewiz I haven't yet added any specific tests. Although I have done some testing and it does pass your reported issue. I would be very happy if you wanted to contribute a bunch of XQuery tests for this (before we merge the PR)?

There is a question over whether this PR should go into eXist-db 3.6.0 or 4.0.0, as there are some method signature changes, which whilst I consider them to be very much internal, perhaps others might complain/disagree; so we need to discuss that @wolfgangmm.

joewiz commented:
@adamretter For **numeric**, would this work? http://icu-project.org/apiref/icu4j/com/ibm/icu/text/RuleBasedCollator.html#getNumericCollation--. This appears to be what BaseX uses; see https://github.com/BaseXdb/basex/blob/098629e0150eb865d61c1c348ff955b43c0014ba/basex-core/src/main/java/org/basex/query/util/collation/UCAOptions.java#L125-L127.

Scrolling up in on that most link also shows BaseX's approach to `alternate` (i.e., [`setAlternateHandlingShifted`](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/RuleBasedCollator.html#setAlternateHandlingShifted-boolean-)), `fallback` (see [fallback parsing](https://github.com/BaseXdb/basex/blob/098629e0150eb865d61c1c348ff955b43c0014ba/basex-core/src/main/java/org/basex/query/util/collation/BaseXCollationOptions.java#L74-L111)), `backwards` ([`setFrenchCollation`](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/RuleBasedCollator.html#setFrenchCollation-boolean-)), `caseLevel` ([`setCaseLevel`](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/RuleBasedCollator.html#setCaseLevel-boolean-)), and `caseFirst` ([`setUpperCaseFirst`](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/RuleBasedCollator.html#setUpperCaseFirst-boolean-) and [`setLowerCaseFirst`](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/RuleBasedCollator.html#setLowerCaseFirst-boolean-)).

Already, this PR helps eXist pass a number of the `fn:compare` tests, namely 017-030. https://github.com/LeoWoerteler/QT3TS/blob/master/fn/compare.xml#L552-L690 - which is a huge step forward. (For anyone wondering how collations are useful, see the list at http://userguide.icu-project.org/collation; search for "here are some of the ways languages vary in ordering strings" and see the listing there.) 

I plan either adapt the QT3 tests to XQSuite or work toward direct integration of the QT3 tests into eXist.

Lastly, incorporating ICU4J also opens up other exciting uses of this library, including the addition of a "transliterate" function that could, for example, transliterate between Latin and Syriac, or between traditional and simplified Chinese, etc. See http://userguide.icu-project.org/transforms/general#TOC-ICU-Transliterators.

adamretter commented:
@joewiz Excellent detective work sir! I have now implemented everything apart from `fallback` as I am not clear yet from studying the BaseX code, quite how that should work, in fact I am not certain that they have it right.

dizzzz commented:
how are things performance wise?

dizzzz commented on .idea/libraries/extensions.xml:
```diff
@@ -8,7 +8,9 @@
     <JAVADOC />
-    <SOURCES />
+    <SOURCES>
+      <root url="jar://$USER_HOME$/.ideaLibSources/lucene-analyzers-common-4.10.4-sources.jar!/" />
```
not sure if this should be in the PR

adamretter replied:
@dizzzz It won't hurt. But I can remove it if you prefer?

joewiz commented:
@dizzzz Re: performance, see http://site.icu-project.org/home/why-use-icu4j#TOC-Performance-Size, especially:

> - Instantiation times are comparable
>     - Common instantiate and reuse model
>     - ICU4J and Java both use caches to limit impact
> - Collation performance many times faster
>     - sorting: 2 to 20 times faster
>     - sort key generation: 1.5 to 4 times faster
>     - sort key length: 2/3 to 1/4 the length of Java sort keys
> - Property access much faster (isLetter, isWhitespace,...)
>
> ICU4J is for you if
> - you need better performance

dizzzz commented:
code looks good, great. I don't see a problem for merging into 3.6 but my arguments might not in line with the agreements :)

joewiz commented:
@adamretter Wonderful!

Now we pass all UCA tests for `fn:compare` except [compare-041](https://github.com/LeoWoerteler/QT3TS/blob/master/fn/compare.xml#L804-L814), which returns this:

```xml
<test-case name="compare-041" covers="uca-collation">
    <created by="Michael Kay" on="2016-09-23"/>
    <modified by="O'Neil Delpratt" on="2016-10-25"
        change="Bug #29931: Added dependency advanced-uca-fallback"/>
    <dependency type="feature" value="advanced-uca-fallback"/>
    <dependency type="spec" value="XP31 XQ31"/>
    <test>fn:compare("database", "data base",
        "http://www.w3.org/2013/collation/UCA?lang=en;alternate=blanked;strength=quaternary") =
        0</test>
    <result>
        <assert-true/>
    </result>
    <wrong>
        <query>declare base-uri "file:/Users/joe/workspace/QT3TS/fn/"; fn:compare("database",
            "data base",
            "http://www.w3.org/2013/collation/UCA?lang=en;alternate=blanked;strength=quaternary")
            = 0</query>
        <out>
            <result>false()</result>
            <errors>
                <error>Query doesn't evaluate to 'true'</error>
            </errors>
        </out>
    </wrong>
</test-case>
```

Oddly, 041 isn't even listed on [the current QT3 results page](https://dev.w3.org/2011/QT3-test-suite/ReportingResults31/testSets/fn-compare.html); but since it lists BaseX 8.5 beta, its certainly out of date.

From `fn-contains`, we have these failures - which seem to relate to `numeric=yes` and `alternate=blanked`:

```xml
<failures xmlns="http://www.w3.org/2010/09/qt-fots-catalog">
    <test-case name="fn-contains-32" covers="uca-collation">
        <created by="Michael Kay" on="2016-09-23"/>
        <modified by="O'Neil Delpratt" on="2016-10-25"
            change="Bug #29931: Added dependency advanced-uca-fallback"/>
        <dependency type="feature" value="advanced-uca-fallback"/>
        <dependency type="spec" value="XP31 XQ31"/>
        <test>fn:contains("Chapter-001", "Chapter-1",
            "http://www.w3.org/2013/collation/UCA?lang=en;numeric=yes")</test>
        <result>
            <any-of>
                <assert-true/>
                <error code="FOCH0004"/>
            </any-of>
        </result>
        <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/fn/";
                fn:contains("Chapter-001", "Chapter-1",
                "http://www.w3.org/2013/collation/UCA?lang=en;numeric=yes")</query>
            <out>
                <result>Error: [ java:java.lang.UnsupportedOperationException ] Numeric collation is
                    not supported by StringSearch</result>
                <errors>
                    <error>Wrong error code [java:java.lang.UnsupportedOperationException] (Numeric
                        collation is not supported by StringSearch), expected: [FOCH0004]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="fn-contains-33" covers="uca-collation">
        <created by="Michael Kay" on="2016-09-23"/>
        <modified by="O'Neil Delpratt" on="2016-10-25"
            change="Bug #29931: Added dependency advanced-uca-fallback"/>
        <dependency type="feature" value="advanced-uca-fallback"/>
        <dependency type="spec" value="XP31 XQ31"/>
        <test>fn:contains("Chapter-100", "Chapter-10",
            "http://www.w3.org/2013/collation/UCA?lang=en;numeric=yes")</test>
        <result>
            <any-of>
                <assert-false/>
                <error code="FOCH0004"/>
            </any-of>
        </result>
        <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/fn/";
                fn:contains("Chapter-100", "Chapter-10",
                "http://www.w3.org/2013/collation/UCA?lang=en;numeric=yes")</query>
            <out>
                <result>Error: [ java:java.lang.UnsupportedOperationException ] Numeric collation is
                    not supported by StringSearch</result>
                <errors>
                    <error>Wrong error code [java:java.lang.UnsupportedOperationException] (Numeric
                        collation is not supported by StringSearch), expected: [FOCH0004]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="fn-contains-37" covers="uca-collation">
        <created by="Michael Kay" on="2016-09-23"/>
        <dependency type="spec" value="XP31 XQ31"/>
        <test>fn:contains("---", "-",
            "http://www.w3.org/2013/collation/UCA?lang=en;alternate=blanked")</test>
        <result>
            <assert-true/>
        </result>
        <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/fn/"; fn:contains("---", "-",
                "http://www.w3.org/2013/collation/UCA?lang=en;alternate=blanked")</query>
            <out>
                <result>false()</result>
                <errors>
                    <error>Query doesn't evaluate to 'true'</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="fn-contains-38" covers="uca-collation">
        <created by="Michael Kay" on="2016-09-23"/>
        <modified by="O'Neil Delpratt" on="2016-10-25"
            change="Bug #29931: Added dependency advanced-uca-fallback"/>
        <dependency type="feature" value="advanced-uca-fallback"/>
        <dependency type="spec" value="XP31 XQ31"/>
        <test>fn:contains("-", "----",
            "http://www.w3.org/2013/collation/UCA?lang=en;alternate=blanked")</test>
        <result>
            <assert-true/>
        </result>
        <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/fn/"; fn:contains("-", "----",
                "http://www.w3.org/2013/collation/UCA?lang=en;alternate=blanked")</query>
            <out>
                <result>false()</result>
                <errors>
                    <error>Query doesn't evaluate to 'true'</error>
                </errors>
            </out>
        </wrong>
    </test-case>
</failures>
```

I also checked `fn-starts-with`, `fn-ends-with`, `fn-substring-before`, `fn:substring-after` and found the same set of errors as here - so I won't paste them in. The tests for `fn-deep-equal` didn't have any UCA-related failures. 

I also ran the `misc-UCAcollection` and got quite a few failures relating to the parameters with value `unknown`: `keyword=unknown`, `strength=unknown`, `alternate=unknown`, `backwards=unknown`, `normalization=unknown`, `caseLevel=unknown`, `caseFirst=unknown`, `numeric=unknown`, and `version=unknown`. Also there seems to be some issues relating to values for the `version` parameter. I'm not sure why `unknown` is tested, since the spec doesn't say anything about this being an accepted value (https://www.w3.org/TR/xpath-functions-31/#uca-collations). I'm just passing these on for completeness.

```xml
<failures xmlns="http://www.w3.org/2010/09/qt-fots-catalog">
    <test-case name="UCA-collation-002">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;keyword=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;keyword=unknown'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-003">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;strength=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;strength=unknown'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;strength=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown</result>
                <errors>
                    <error>Wrong error code [java:org.exist.xquery.XPathException] (exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;strength=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown), expected: [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-004">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null</result>
                <errors>
                    <error>Wrong error code [java:org.exist.xquery.XPathException] (exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null), expected: [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-005">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;backwards=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;backwards=unknown'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-006">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;normalization=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;normalization=unknown'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-007">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;caseLevel=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseLevel=unknown'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-008">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown</result>
                <errors>
                    <error>Wrong error code [java:org.exist.xquery.XPathException] (exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown), expected: [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-010">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'fallback=no;numeric=unknown'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'fallback=no;numeric=unknown'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-012">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','def',concat($collation,'strength=unknown'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'strength=unknown'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'strength=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown</result>
                <errors>
                    <error>Expected result, found error: [java:org.exist.xquery.XPathException] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'strength=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'strength=unknown')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: unknown</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-013">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','def',concat($collation,'alternate=unknown'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'alternate=unknown'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'alternate=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null</result>
                <errors>
                    <error>Expected result, found error: [java:org.exist.xquery.XPathException] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'alternate=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'alternate=unknown')). Collation alternate should be either 'non-ignorable', 'shifted' or 'blanked', but received: null</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-017">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','def',concat($collation,'caseFirst=unknown'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'caseFirst=unknown'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'caseFirst=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown</result>
                <errors>
                    <error>Expected result, found error: [java:org.exist.xquery.XPathException] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'caseFirst=unknown'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'caseFirst=unknown')). Collation case first should be either 'upper' or 'lower', but received: unknown</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-020">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','def',concat($collation,'version=5.0'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=5.0'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=5.0')). Requested UCA Collation version: 5.0, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=5.0'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=5.0')). Requested UCA Collation version: 5.0, however eXist-db only has ICU UCA: 137.51.0.0</result>
                <errors>
                    <error>Expected result, found error: [java:org.exist.xquery.XPathException] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=5.0')). Requested UCA Collation version: 5.0, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=5.0'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=5.0')). Requested UCA Collation version: 5.0, however eXist-db only has ICU UCA: 137.51.0.0</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-021">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','def',concat($collation,'version=6.0'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=6.0'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=6.0')). Requested UCA Collation version: 6.0, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=6.0'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=6.0')). Requested UCA Collation version: 6.0, however eXist-db only has ICU UCA: 137.51.0.0</result>
                <errors>
                    <error>Expected result, found error: [java:org.exist.xquery.XPathException] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=6.0')). Requested UCA Collation version: 6.0, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=6.0'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','def',concat($collation,'version=6.0')). Requested UCA Collation version: 6.0, however eXist-db only has ICU UCA: 137.51.0.0</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-022">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="O'Neil Delpratt" on="2014-12-16" change="Bug fix relating #27180: Change error code"/>
      <modified by="O'Neil Delpratt" on="2015-07-30" change="Bug fix 28944: Change 'aBC' to 'xyz'"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','xyz',concat($collation,'version=7.0'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=7.0'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=7.0')). Requested UCA Collation version: 7.0, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=7.0'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=7.0')). Requested UCA Collation version: 7.0, however eXist-db only has ICU UCA: 137.51.0.0</result>
                <errors>
                    <error>Expected result, found error: [java:org.exist.xquery.XPathException] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=7.0')). Requested UCA Collation version: 7.0, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=7.0'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=7.0')). Requested UCA Collation version: 7.0, however eXist-db only has ICU UCA: 137.51.0.0</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-022a">
      
      <created by="O'Neil Delpratt" on="2014-12-18"/>
      <modified by="Christian Gruen" on="2015-01-27" change="minor version number limited to smaller value"/>
      <modified by="O'Neil Delpratt" on="2015-07-30" change="Bug fix 28944: Change aBC to xyz"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','xyz',concat($collation,'version=1.255'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=1.255'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=1.255')). Requested UCA Collation version: 1.255, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=1.255'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=1.255')). Requested UCA Collation version: 1.255, however eXist-db only has ICU UCA: 137.51.0.0</result>
                <errors>
                    <error>Expected result, found error: [java:org.exist.xquery.XPathException] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=1.255')). Requested UCA Collation version: 1.255, however eXist-db only has ICU UCA: 137.51.0.0 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=1.255'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=1.255')). Requested UCA Collation version: 1.255, however eXist-db only has ICU UCA: 137.51.0.0</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-023">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-07-13" change="should use fallback. (But see bug 28944)"/>
      <modified by="O'Neil Delpratt" on="2015-07-30" change="Bug fix 28944: Change aBC to xyz"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: document dependency on simple fallback behaviour"/>
      <environment ref="UCA-collation"/>
      <dependency type="feature" value="simple-uca-fallback"/>
      <test>compare('abc','xyz',concat($collation,'version=unknown'))</test>
      <result>
         <assert-eq>-1</assert-eq>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=unknown'))</query>
            <out>
                <result>Error: [ java:java.lang.IllegalArgumentException ] Invalid version number: Version number may be negative or greater than 255. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=unknown')). Invalid version number: Version number may be negative or greater than 255</result>
                <errors>
                    <error>Expected result, found error: [java:java.lang.IllegalArgumentException] Invalid version number: Version number may be negative or greater than 255. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','xyz',concat($collation,'version=unknown')). Invalid version number: Version number may be negative or greater than 255</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-collation-024">
      
      <created by="Michael Kay" on="2015-07-13"/>
      <environment ref="UCA-collation"/>
      <test>compare('abc','aBC',concat($collation,'version=96.5;fallback=no'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?';
compare('abc','aBC',concat($collation,'version=96.5;fallback=no'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-reorder-codes-001">
      
      <created by="John Lumley" on="2014-05-02"/>
      <environment ref="UCA-collation.en.primary"/>
      <test>compare('.123','123',concat($collation,'fallback=no;reorder=Z,digit'))</test>
      <result>
         <error code="FOCH0002"/>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;strength=primary;';
compare('.123','123',concat($collation,'fallback=no;reorder=Z,digit'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-reorder-codes-005">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: disallow fallback behaviour"/>
      <environment ref="UCA-collation.en.primary"/>
      <test>compare('.123','123',concat($collation,'reorder=digit,punct;fallback=no'))</test>
      <result>
         <any-of>
            <assert-eq>1</assert-eq>
            <error code="FOCH0002"/>
         </any-of>       
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;strength=primary;';
compare('.123','123',concat($collation,'reorder=digit,punct;fallback=no'))</query>
            <out>
                <result>-1</result>
                <errors>
                    <error>Result doesn't match expected item '1'.</error>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-reorder-codes-010">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: disallow fallback behaviour"/>
      <environment ref="UCA-collation.en.primary"/>
      <test>compare('a123','123',concat($collation,'reorder=Latn,digit;fallback=no'))</test>
      <result>
         <any-of>
            <assert-eq>-1</assert-eq>
            <error code="FOCH0002"/>
         </any-of>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;strength=primary;';
compare('a123','123',concat($collation,'reorder=Latn,digit;fallback=no'))</query>
            <out>
                <result>1</result>
                <errors>
                    <error>Result doesn't match expected item '-1'.</error>
                    <error>Expected Error [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-params-010">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: disallow fallback behaviour"/>
      <environment ref="UCA-collation.en"/>
      <test>compare('abc','aBC',concat($collation,'strength=5;fallback=no'))</test>
      <result>
         <any-of>
            <assert-eq>-1</assert-eq>
            <error code="FOCH0002"/>
         </any-of>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('abc','aBC',concat($collation,'strength=5;fallback=no'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('abc','aBC',concat($collation,'strength=5;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('abc','aBC',concat($collation,'strength=5;fallback=no'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('abc','aBC',concat($collation,'strength=5;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5</result>
                <errors>
                    <error>Wrong error code [java:org.exist.xquery.XPathException] (exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('abc','aBC',concat($collation,'strength=5;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('abc','aBC',concat($collation,'strength=5;fallback=no'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('abc','aBC',concat($collation,'strength=5;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5), expected: [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
    <test-case name="UCA-params-015">
      
      <created by="John Lumley" on="2014-05-02"/>
      <modified by="Michael Kay" on="2015-12-02" change="Bug 29308: disallow fallback behaviour"/>
      <environment ref="UCA-collation.en"/>
      <test>compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no'))</test>
      <result>
         <any-of>
            <assert-eq>-1</assert-eq>
            <error code="FOCH0002"/>
         </any-of>
      </result>
   <wrong>
            <query>declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no'))</query>
            <out>
                <result>Error: [ java:org.exist.xquery.XPathException ] exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5</result>
                <errors>
                    <error>Wrong error code [java:org.exist.xquery.XPathException] (exerr:ERROR Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5 [at line 57, column 3, source: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no'))]
In function:
	local:eval(xs:string) [129:30:String]. Error while evaluating expression: declare base-uri "file:/Users/joe/workspace/QT3TS/misc/";
declare variable $collation := 'http://www.w3.org/2013/collation/UCA?lang=en;';
compare('a-bc','abc',concat($collation,'strength=5;alternate=shifted;fallback=no')). eXist-db only supports Collation strengths of 'identical', 'primary', 'secondary', 'tertiary' or 'quaternary', requested: 5), expected: [FOCH0002]</error>
                </errors>
            </out>
        </wrong>
    </test-case>
</failures>
```

adamretter commented:
@joewiz Some further improvements for you:

1. I fixed our implementation so that `fn-contains-32` and `fn-contains-33` now pass.
2. I can't seen to figure out how to pass `fn-contains-37`.
3. The XQTS test for `fn-contains-38` you pasted above seems wrong or outdated?!? Both Saxon and BaseX return `false()` for that, as does eXist-db.
4. I think I have fixed the issue where "unknown" was reporting the wrong error code.

wolfgangmm commented on src/org/exist/start/start.config:
```diff
@@ -62,7 +62,10 @@ lib/core/xmlrpc-server-%latest%.jar                always
 lib/core/clj-ds-%latest%.jar                       always
 lib/core/cglib-nodep-%latest%.jar                  always
 lib/core/j8fu-%latest%.jar			   always
+lib/core/icu4j-%latest%.jar                        always
```
When testing this branch, the library was not found unless I used the explicit version '59_1' instead of %latest%. Maybe it fails because of the "_1". Not sure why I got it. I made a `clean-all` before testing.

joewiz replied:
Is it possible you had an old version of icu4j in lib/user? In my testing I had encountered a `java.lang.NoSuchMethodError`, and when I asked Adam he said: "do you have an older icu jar kicking around in lib/user or somewhere?" I found I had icu4j-4_8_1_1.jar in lib/user (dated Aug 23, 2011!). Must've been there from a previous "download-icu4j" build step. Doing a `clean-all` didn't help. I had to manually delete the old jar file.

adamretter replied:
@wolfgangmm Can you check what @joewiz  suggested please?

wolfgangmm replied:
No, I did not have an old icu. I was testing on a fresh checkout and it definitely fails to load `icu4j-59_1.jar`.

adamretter replied:
@wolfgangmm Okay I see there was an issue when starting eXist from `org.exist.start.Main`, I have now fixed that whe

joewiz commented:
@adamretter I recall you saying that you wanted to drop the `download-icu4j` build target. Do you want to include that in this PR?

adamretter commented:
@joewiz It's already in the PR I believe.
